### PR TITLE
[SPARK-39444][SQL] Add OptimizeSubqueries into nonExcludableRules list

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This PR adds `OptimizeSubqueries` rule into nonExcludableRules list.

### Why are the changes needed?

It will throw exception if user `set spark.sql.optimizer.excludedRules=org.apache.spark.sql.catalyst.optimizer.Optimizer$OptimizeSubqueries` before running this query:
```sql
WITH tmp AS (
  SELECT id FROM range(2)
  INTERSECT
  SELECT id FROM range(4)
)

SELECT id FROM range(5) WHERE id > (SELECT max(id) FROM x)
```
Exception:
```
logical intersect  operator should have been replaced by semi-join in the optimizer
java.lang.IllegalStateException: logical intersect  operator should have been replaced by semi-join in the optimizer
```

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test.

### CODE REVIEW

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
cc @maryannxue and @allisonwang-db FYI

**Review by allisonwang-db — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 1]
```diff
@@ -4456,6 +4456,24 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
         """.stripMargin),
       Seq(Row(2), Row(1)))
   }
+
+  test("SPARK-39444: Add OptimizeSubqueries into nonExcludableRules list") {
+    withSQLConf(
+      SQLConf.OPTIMIZER_EXCLUDED_RULES.key ->
+        "org.apache.spark.sql.catalyst.optimizer.Optimizer$OptimizeSubqueries") {
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:4463]: nit but can we use a class reference instead of a string?

Comment (HyukjinKwon) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:4463]: and let's match the test place with https://github.com/apache/spark/pull/36847

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:4463]: +1 for @HyukjinKwon 's comment.

Comment (wangyum) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:4463]: fixed.

File: sql/core/src/test/resources/sql-tests/inputs/non-excludable-rule.sql [hunk 2]
```diff
@@ -4,3 +4,12 @@ SELECT
   (SELECT min(id) FROM range(10)),
   (SELECT sum(id) FROM range(10)),
   (SELECT count(distinct id) FROM range(10));
+
+-- SPARK-39444
+SET spark.sql.optimizer.excludedRules=org.apache.spark.sql.catalyst.optimizer.Optimizer$OptimizeSubqueries;
```
Comment (dongjoon-hyun) [inline sql/core/src/test/resources/sql-tests/inputs/non-excludable-rule.sql:9]: Although it was working, I didn't realize that we allow to exclude rules whose prefix is `Optimizer$...`. Is it legitimate? I'm just curious if we need to ban all exclusions for ``Optimizer$...` rules.

Comment (wangyum) [reply sql/core/src/test/resources/sql-tests/inputs/non-excludable-rule.sql:9]: Another rule is: `org.apache.spark.sql.catalyst.optimizer.Optimizer$UpdateCTERelationStats`.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**